### PR TITLE
Reinitialize segments after welford translation

### LINKED ISSUE #2348: Disjointness of segmented fusion may be invalidated due to welford translation
Translation of welford exprs reuses existing exprs if found:

https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/fusion_segmenter.cpp#L2013-L2036

This however pulls in expressions that may belong to different segmented groups than the translated group. One place we do the translation is at the beginning of the segmentation where we have initial segmented groups, i.e., each group consists of a single expr. If the translation is done at this point, reusing would mean pulling in exprs to the welford group, but those exprs still exist in their original initial groups, so the disjointness gets invalidated.

To fix the issue, we should either just stop reusing or reinitialize segmented groups. 

I found this happened with some of the normalization tests such as `TestCudaFuser.test_norm_half`.

### PR DESCRIPTION
**Pull request by naoyam**
Closes #2348

### CODE REVIEW

File: third_party/nvfuser/test/test_gpu_outer_reduction.cpp [hunk 1]
```diff
@@ -411,7 +411,8 @@ void scheduleNormalization(Fusion& fusion, const OuterReductionParams& params) {
           input_tv->toString());
       auto cast_tv = dynamic_cast<TensorView*>(cast_expr->out());
       TORCH_CHECK(cast_tv != nullptr);
-      for (auto use : cast_tv->uses()) {
+      auto cast_tv_use_exprs = cast_tv->uses();
+      for (auto use : cast_tv_use_exprs) {
```
Comment (naoyam) [inline third_party/nvfuser/test/test_gpu_outer_reduction.cpp:415]: Unrelated bug fix

File: third_party/nvfuser/csrc/fusion_segmenter.cpp [hunk 2]
```diff
@@ -2647,15 +2642,16 @@ SegmentCandidateFinder::SegmentCandidateFinder(
   findSegments();
 }
 
-void SegmentCandidateFinder::findSegments() {
-  FUSER_PERF_SCOPE("Finding valid fusion segment solutions");
+void SegmentCandidateFinder::buildInitialSegments() {
```
Comment (naoyam) [inline third_party/nvfuser/csrc/fusion_segmenter.cpp:2631]: Factored out the part to create the initial graph so that it can be redone if necessary

File: third_party/nvfuser/csrc/fusion_segmenter.cpp [hunk 3]
```diff
@@ -1831,12 +1831,7 @@ TranslateApplicableWelford::TranslateApplicableWelford(
   // replaced by the translation.
   for (auto welford : welford_to_translate) {
     translateSingleWelford(welford);
-  }
-
-  for (auto translated_group : translated_groups) {
-    // Update heuristics and expr list of translated groups
-    translated_group->heuristic_ = ScheduleHeuristic::Persistent;
-    translated_group->resetExprList();
```
Comment (naoyam) [inline third_party/nvfuser/csrc/fusion_segmenter.cpp:1839]: No longer makes any sense as it will be regrouped.

File: third_party/nvfuser/csrc/fusion_segmenter.cpp [hunk 4]
```diff
@@ -3145,6 +3172,66 @@ void SegmentedFusion::setCachedHeuristicDataFor(
   heuristic_summary_cache_[group] = std::move(data);
 }
 
+void SegmentedFusion::validate(bool require_disjoint) const {
```
Comment (naoyam) [inline third_party/nvfuser/csrc/fusion_segmenter.cpp]: Some sanity checks that makes me feel more confident. Had some difficulties in debugging. These checks would have been helpful.

File: third_party/nvfuser/csrc/fusion_segmenter.cpp [hunk 5]
```diff
@@ -2690,6 +2686,7 @@ void SegmentCandidateFinder::findSegments() {
     while (!to_visit.empty()) {
       auto expr = to_visit.front();
       to_visit.pop_front();
+      // TODO: fix here. See issue #2351
```
Comment (naoyam) [inline third_party/nvfuser/csrc/fusion_segmenter.cpp:2675]: Will work on it as a separate PR

File: third_party/nvfuser/csrc/fusion_segmenter.cpp [hunk 6]
```diff
@@ -3145,6 +3172,66 @@ void SegmentedFusion::setCachedHeuristicDataFor(
...
+void SegmentedFusion::validate(bool require_disjoint) const {
+  // Make sure the groups are a DAG
+  std::unordered_set<SegmentedGroup*> input_groups;
+  std::copy_if(
+      groups().begin(),
+      groups().end(),
+      std::inserter(input_groups, input_groups.end()),
+      [](SegmentedGroup* group) { return group->isInputGroup(); });
+
+  std::unordered_set<SegmentedEdge*> visited_edges;
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/fusion_segmenter.cpp]: Are you emplacing to this set? I don't see it.

File: third_party/nvfuser/csrc/fusion_segmenter.cpp [hunk 7]
```diff
@@ -3145,6 +3172,66 @@ void SegmentedFusion::setCachedHeuristicDataFor(
...
+
+  std::deque<SegmentedGroup*> to_visit(
+      {input_groups.begin(), input_groups.end()});
+
+  while (!to_visit.empty()) {
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/fusion_segmenter.cpp]: I'm not sure what the requirement is that's being checked here. I don't know what the sufficient check is for checking for a correct DAG. I guess I would attempt to traverse it and find if I get stuck in the traversal, but maybe the check you're trying here is simpler?

**Review by csarofeen — COMMENTED** [review]
Just one question on the validation, otherwise LGTM.

File: third_party/nvfuser/csrc/fusion_segmenter.cpp (hunk 6 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp]: Oops... Will fix it

File: third_party/nvfuser/csrc/fusion_segmenter.cpp (hunk 7 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp]: There should have been `visited_edges.insert(consumer_edge)`.

Comment (csarofeen) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp]: Don't you need to make sure all producer edges were visited before traversing a group? Otherwise I think you'd have to traverse starting at one input at a time no?

Comment (naoyam) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp]: I changed it to do DFS traversals in a similar way as `IterVisitor::traverseBetween`.

**Comment by naoyam** [comment]
@csarofeen I believe the DAG detection is fixed, though I don't have any test case. All tests are green.

**Review by csarofeen — APPROVED** [review]
LGTM
